core/types/calls.cc

```diff
@@ -1532,6 +1532,10 @@ TypePtr getMethodArguments(const GlobalState &gs, ClassOrModuleRef klass, NameRe
         if (arg.flags.isBlock) {
             continue;
         }
+        if (arg.type == nullptr) {
+            args.emplace_back(core::Types::untyped(gs, method));
+            continue;
+        }
```

Comment:
Naively, I would assume that this PR fixes the type of a block argument that wasn't a proc-typed thing...but it looks like we handle block arguments in the conditional just above this.

Is this instead fixing the behavior in the testcase where we want to know what the type of `x` is for `CallTakesBlock#call`?  Or is is fixing the case where we're trying to see what the type of `x` is in the block for the call to `example`?

Reply:
`getMethodArguments` is passed a `NameRef name`, which is maybe a bit more generic than it needs to be, because it is always passed the `core::Names::call()` if you trace the call stack all the way up:

```cpp
auto params = procType.getCallArguments(ctx, core::Names::call());
```

What happens is that we look up the method on that type with the name `call` (possibly unwrapping things like `T.any` and `T.all` to get there). Once we've found the `call` method, we look at it's args and construct a tuple type with the type for all the args, so that we can have a `LoadYieldParams` instruction in the CFG that is typed as that tuple:

```ruby
example do |x|
  T.reveal_type(x) # error: `T.untyped`
end
```

becomes

```
<args> : [T.untyped] = LoadYieldParams(example)
x : T.untyped = <args>.[](0)
```

(if there had been more arguments in the block, like `do |x, y, z|`, we would have had things like `<args>.[](1)`, etc.)

So in this case, `example` says its block (`procType`) is `CallTakesBlock`, then we look up the `call` method on that type, and want to construct a 1-tuple representing the types of the arguments that `CallTakesBlock#call` takes, which should be `[T.untyped]`.

But the problem was that if there is no sig on `call`, then `arg.type` is `nullptr`, which meant we constructed `[nullptr]`, which causes things to crash.

> Is this instead fixing the behavior in the testcase where we want to know what the type of `x` is for `CallTakesBlock#call`? Or is is fixing the case where we're trying to see what the type of `x` is in the block for the call to `example`?

It's kind of both? We need to get the type of `x` off the symbol for `CallTakesBlock#call` to see what the type of `x` is in the block for the call to `example`.

I guess maybe the best answer to your question is the latter, because this `getMethodArguments` function for sure runs while infer is running to typecheck the call to `example do...end`, not to typecheck the body of `CallTakesBlock#call`.